pkg/logbridge/logbridge.go

```diff
@@ -0,0 +1,116 @@
+package logbridge
+
+import (
+	"bufio"
+	"io"
+	"time"
+
+	"github.com/square/p2/pkg/logging"
+)
+
+// Copy implements a buffered copy operation between dest and src.
+// It returns the number of dropped messages as a result of insufficient
+// capacity
+// The caller is responsible for closing & flushing the reader and writer.
+func Copy(dest io.Writer, src io.Reader, capacity int, logger logging.Logger) {
+	lines := make(chan []byte, capacity)
+	droppedMessages := make(chan []byte, capacity)
+	go func(dest io.Writer, src io.Reader, lines chan []byte) {
+		defer close(lines)
+
+		scanner := bufio.NewScanner(src)
+		var line []byte
+		for scanner.Scan() {
+			line = scanner.Bytes() // consume a line regardless of the state of the writer
+			select {
+			case lines <- line:
+			default:
+				select {
+				case droppedMessages <- line:
+				default:
+				}
+				logger.WithField("dropped line", line).Errorln("Dropped was dropped due to full capacity. If this occurs frequently, consider increasing the capacity of this logbridge.")
+			}
+		}
+		if err := scanner.Err(); err != nil {
+			logger.WithError(err).Errorln("Encountered error while reading from src. Proceeding.")
+		}
+	}(dest, src, lines)
+
+	go func(droppedMessages chan []byte) {
```

Comment:
broadly speaking, I don't see a lot of value in handling dropped messages via a separate goroutine. If counting the number of dropped messages is important, you can implement it locklessly via an int from sync/atomic. The thing about handling dropped messages via a channel is that the dropped messages channel might itself be so backed up that you have to drop the message another way. And if you're going to have another way to drop messages without using channels, why not just use that way all the time?